mesonbuild/coredata.py

```diff
@@ -704,12 +704,11 @@ def replace(v):
 
         if key.name == 'buildtype':
             dirty |= self._set_others_from_buildtype(value)
-        elif key.name in {'wrap_mode', 'force_fallback_for'}:
+        elif changed and key.name in {'wrap_mode', 'force_fallback_for'}:
             # We could have the system dependency cached for a dependency that
             # is now forced to use subproject fallback. We probably could have
```

Comment:
This seems like it should be incredibly wrong. Why do we care if the current value has changed since the last time we set it? We should care what the current value is. 

Actually why do we care about these settings at all? If we do care, then the cache is inherently broken. The purpose of this is to preserve expensive values that require shelling out a lot to pkg-config / cmake, not to preserve the most recent resolution strategy. We need to be caching this all the time (regardless of wrap mode) and then having every dependency lookup re-evaluate methods, fallback state, and so on -- and only looking up the cache once it's clear that we would have used that cached dependency anyway.

Reply:
> If we do care, then the cache is inherently broken.

The understatement of the year :D

That's just a cheap and easy way to deal with cache invalidation, but feel free to take the hard path if you want.

Reply:
Actually this looks pretty easy, hold on a minute...

Basically, we have most of the relevant handling already. The trick is that if we have a cached pkg-config dep but also forced the fallback we will go ahead and look it up in the cache, which was weird.